### ISSUE TITLE
Recommended way to integrate with existing JS widgets/components?

### ISSUE THREAD

**Issue by saurabhnanda** [dmjio/miso#222]
What would be the recommended way to integrate existing JS widgets/components in a Miso app? say, the jQuery datepicker, or Bootstrap menus, or jQuery fullCalendar, etc?

**Comment by dmjio** [comment]
After creating a view, it's possible to put an `id` on it, i.e. `div_ [ id_ "datepicker" ] []`.
And then make an ffi call to retrieve it.
```
foreign import javascript unsafe "$r = document.getElementById($1);"
   getElementById :: JSString -> IO JSVal
```

and then `ffi` into your `jQuery` to do something like:
```haskell
update InitialAction m = m <# do
  jsval <- getElementById "datepicker"
  initDatePicker jsval
  pure NoOp
```

Note, you'd have to take care to hide it (`visibility = hidden` -- in html or css) if you navigate away, otherwise it will be blown away (since the diffing algorithm isn't aware of child nodes on `datepicker`). In worst case loss of synchrony can occur between the virtual dom and real dom. Otherwise, you can probably re-init it, when navigating to a new page.

**Comment by saurabhnanda** [comment]
> Note, you'd have to take care to hide it (visibility = hidden -- in html or css), or ensure you never touch it. If you navigate away, it will be blown away (since the diffing algorithm isn't aware of child nodes on datepicker). In worst case loss of synchrony can occur between the virtual dom and real dom. Otherwise, you can probably re-init it, when navigating to a new page.

Is there a safer way to get this done? Reusing existing JS components seems critical to put Miso into real-life use. Else a lot of time would be spent in any project trying re-implement common UI widgets (datepicker, calendar, auto-complete, sliders, select2, etc) in Miso.

How does the React/Elm world deal with this problem? I'm certain they would have been in a similar phase at the beginning of their projects.

**Comment by saurabhnanda** [comment]
Are there any DOM lifecycle events exposed by Miso that a component can hook into? eg.

* Component about to be removed by the vdom diffing engine ==> execute the jquery component's `destroy` method in this hook
* Component added by the vdom diffing engine ==> execute the jquery component's `init` method in this hook

**Comment by dmjio** [comment]
> Is there a safer way to get this done? 

I'd say not really, in Elm you'd have to use ports, and even then, if you're adding / removing DOM nodes, you're always at the risk of losing synchrony between DOM and virtual DOM. `jQuery` is kind of orthogonal to virtual DOM. 

I would leave all DOM manipulation to virtual DOM, if you can help it. Or in cases like canvas, hide it when not using it and don't remove it from the virtual-dom.

Although, if you have a template like
```haskell
div_ [ id_ "datepicker" ] []
```
And then you, from javascript, initialized a `datepicker` on that node where only child nodes were added, as long as this `div` is never removed from the virtual-dom (in the `view` function), you won't be at risk of losing synchrony (since diff will skip over these child elements).

> Reusing existing JS components seems critical to put Miso into real-life use. 

I've used it in real life w/o existing JS components :)

In other frameworks like react and angular they end up *removing* all jQuery from their UI toolkits (e.g. when wanting to use bootstrap).

https://angular-ui.github.io/bootstrap/
https://react-bootstrap.github.io/
http://elm-bootstrap.info/

Semantic-ui wants to drop it as well.
https://github.com/Semantic-Org/Semantic-UI/issues/1175

> Else a lot of time would be spent in any project trying re-implement common UI widgets (datepicker, calendar, auto-complete, sliders, select2, etc) in Miso.

A lot of pure css-only frameworks provide things like this for you (I've used http://bulma.io/). But yes, something like `miso-bootstrap`  would need to be written, where all the component logic is performed in Haskell. The good news is that if there is something you see in Elm, you can more or less replicate it in Miso, since they are so similar. (The mario and todo-mvc examples were almost copied verbatim).

For datepicker, elm has written their own. https://ellie-app.com/pwFvvCqBgYa1/0
https://github.com/elm-community/elm-datepicker

> Are there any DOM lifecycle events exposed by Miso that a component can hook into? eg.

Not yet, but I've considered adding these. One issue is that we need to keep the view pure, so it's shareable on the server. In general though, miso uses Elm as a good frame of reference for these kinds of things, so I'll see if it is in the pipeline for them.

**Comment by dmjio** [comment]
Also, I'd like to share what Evan says here:
https://guide.elm-lang.org/reuse/

**Comment by saurabhnanda** [comment]
> For datepicker, elm has written their own. https://ellie-app.com/pwFvvCqBgYa1/0
https://github.com/elm-community/elm-datepicker

And therein lies the problem. They seem to be reimplementing years of effort that has already been spent into polishing the UI/UX of a datepicker. Moreover the Elm version still doesn't have feature parity with the jQuery datepicker. (eg. localisation, range restriction, range picking, etc)

> Not yet, but I've considered adding these. One issue is that we need to keep the view pure, so it's shareable on the server. In general though, miso uses Elm as a good frame of reference for these kinds of things, so I'll see if it is in the pipeline for them.

Can it be assumed that code executed during a DOM lifecycle event would be related to in-browser DOM manipulation? Can the server ignore those hook, only to be fired by the in-browser miso app, upon application initialisation?

**Comment by dmjio** [comment]
> And therein lies the problem. They seem to be reimplementing years of effort that has already been spent into polishing the UI/UX of a datepicker. Moreover the Elm version still doesn't have feature parity with the jQuery datepicker. (eg. localisation, range restriction, range picking, etc)

Not saying I don't agree with you here :)

> Can it be assumed that code executed during a DOM lifecycle event would be related to in-browser DOM manipulation? 

Yes, right before the diff, we'd fire the lifecycle event on the node (be it destroy or create).

> Can the server ignore those hook, only to be fired by the in-browser miso app, upon application initialisation?

Yes, on server events are just stubs, no DOM on server, so no events.

**Comment by dmjio** [comment]
> Firing an action via jQuery event listeners shouldnt be a problem in Miso, right?

You could do it by way of a `Subscription`, but I would just use `miso`'s event delegation mechanism.

> How would one wite an efficient viewModel function for something that's wrapping a jQuery component?

I would try to find examples of `Elm` doing this, but you're probably better off not going against the grain and trying to shoe-horn `jQuery` into a paradigm where it isn't best suited. How difficult could it be to recreate a calendar or date-picker in miso? You can even use the `time` library. I'd venture it'd be less trouble to do this, and you'd hit way less bugs.

**Comment by saurabhnanda** [comment]
> How difficult could it be to recreate a calendar or date-picker in miso?

Please compare any Elm library with the corresponding library's in jQuery. Elm lacks polish. It takes a lot of effort to write a polished UI component. 

And it's about time to market. Would we want to be spending time solving domain problems or rewriting UI components? 

I strongly urge you to part ways with Elm here and consider adding top-notch support for integrating with existing JS components.

**Comment by dmjio** [comment]
> Please compare any Elm library with the corresponding library's in jQuery. Elm lacks polish. It takes a lot of effort to write a polished UI component

There's no reason why someone couldn't write a "polished" UI component in Elm or miso. 

> I strongly urge you to part ways with Elm here and consider adding top-notch support for integrating with existing JS components.

You can integrate with other js libs already, and I've outlined how to above. Other users are already doing this.

**Comment by NickSeagull** [comment]
> And it's about time to market. Would we want to be spending time solving domain problems or rewriting UI components?

You might be losing time by "rewriting" UI components, such as a jQuery datepicker (which, in my opinion has horrible aesthetics), but you are not taking into account the other great features, that not only miso, but rather the *language* can offer. "If it compiles, it works" is a sentence that is generally true, which was what Evan (author of Elm) tried to pass from the Haskell community to the JS one.

Heck, we can go even further and make a backend for our webapp, and [forget about WebSockets, HTTP, serialization](https://github.com/theam/haskell-do/blob/master/src/common/HaskellDo/State.hs#L75) and stuff like that by using `atRemote` from other Haskell libraries like [transient](https://github.com/transient-haskell/transient) (it even can deploy microservices!).

What I'm trying to say here, is that even people that are using React.js are moving away from jQuery, because it is not what it is intended to be done. React, Elm, [Pux](http://purescript-pux.org/), Miso... they all share the same ideology, which is to make things more pure to help the developer to reason about them in an easier way.

There's a magnificent [tweet](https://twitter.com/mfeathers/status/29581296216) that says:

> OO makes code understandable by encapsulating moving parts. FP makes code understandable by minimizing moving parts.

And, as much as I love jQuery, it just doesn't fit here. Of course you could just use it using the trick that @dmjio pointed, but it is not what it is natural (you can [write BASIC in Haskell](https://hackage.haskell.org/package/BASIC-0.1.5.0/docs/Language-BASIC.html) but still no one does).

Please, don't take my comments as hostile. On the opposite, I'm very happy that people criticize the Functional point of view, and I just want to help 😁

**Comment by saurabhnanda** [comment]
> You can integrate with other js libs already, and I've outlined how to above. Other users are already doing this.

Can you please point me to some examples? I'm struggling with trying to trigger a Miso event upon a jQuery event (say http://api.jqueryui.com/datepicker/#option-onSelect ).

**Comment by saurabhnanda** [comment]
Is this approach replicable in Miso - http://tech.noredink.com/post/156272288238/picking-dates-with-elm ?

@NickSeagull that article pretty-much sums up what I'm trying to say in this thread:

> A frontend developer sometimes just wants to drop a JavaScript widget somebody else made into their application and move on. Maybe that widget is a slider, or a menu, or a progress bar, or a spinner, or a tab, or a tooltip that points in a cool way.. And sometimes that same frontend developer would like to write their application in Elm.. Should this developer wait to use Elm until the widget they want is rewritten in Elm? Should they rewrite everything that they need?

**Comment by NickSeagull** [comment]
Definitely, I'm not meaning that JS interop shouldn't be done. In fact, the only good thing about JS is it's environment, and we should use the awesome libraries that there are available.

>  I'm struggling with trying to trigger a Miso event upon a jQuery event.

Triggering Miso events based on JS events should be done using [Subscriptions](https://haddocks.haskell-miso.org/Miso-Html.html#t:Sub), like one does in Elm. But because Elm is strict, and Haskell is non-strict, we cannot do this part described in the NoRedInk article:

```javascript
$(function() {
  elmHost = document.getElementById("elm-host")
  var app = Elm.Component.embed(elmHost);

  $.datepicker.setDefaults({
    showOn: "focus",
    onSelect: sendDate,
  });

  app.ports.openDatepicker.subscribe(function() {
    $("#date-input").datepicker().datepicker("show");
  });

  function sendDate (dateString) {
    app.ports.changeDateValue.send(dateString)
  }
});
```

And rather write them from Haskell, by using the GHCJS FFI and the `Foreign.Callback` module:

```haskell
-- | Captures mouse coordinates as they occur and writes them to
-- an event sink
mouseSub :: ((Int,Int) -> action) -> Sub action model
mouseSub f _ = \sink -> do
  windowAddEventListener "mousemove" =<< do
    asyncCallback1 $ \mouseEvent -> do
      Just x <- fromJSVal =<< getProp "clientX" (Object mouseEvent)
      Just y <- fromJSVal =<< getProp "clientY" (Object mouseEvent)
      sink $ f (x,y)
```

( Source from [Miso.Subscription.Mouse](https://haddocks.haskell-miso.org/src/Miso-Subscription-Mouse.html#mouseSub) )

**Comment by saurabhnanda** [comment]
`windowAddEventListener` cannot be used without these hidden modules:

```
/Users/saurabhnanda/projects/miso-js-interop/app/Main.hs:13:1: error:
    Failed to load interface for ‘Miso.FFI’
    it is a hidden module in the package ‘miso-0.4.0.0’
    Use -v to see a list of the files searched for.

/Users/saurabhnanda/projects/miso-js-interop/app/Main.hs:14:1: error:
    Failed to load interface for ‘Miso.Html.Internal’
    it is a hidden module in the package ‘miso-0.4.0.0’
    Use -v to see a list of the files searched for.
```

**Comment by saurabhnanda** [comment]
Here's my WIP code, if anyone is interested - https://github.com/vacationlabs/miso-js-interop/blob/master/app/Main.hs

**Comment by dmjio** [comment]
`windowAddEventListener` is meant to be hidden in `Miso.FFI`.

**Comment by saurabhnanda** [comment]
> windowAddEventListener is meant to be hidden in Miso.FFI.

In that case, what is an alternative to get this js-interop in place?

**Comment by dmjio** [comment]
You'd have to write your own FFI bindings (you can probably just copy the ones from `Miso.FFI`). This will probably be true of most subscriptions. This is a good primer on the GHCJS ffi, https://github.com/ghcjs/ghcjs/blob/master/doc/foreign-function-interface.md

**Comment by levithomason** [comment]
@dmjio quick clarification on:

>Semantic-ui wants to drop it as well.
>Semantic-Org/Semantic-UI#1175

I'm not convinced we need to remove jQuery from the core Semantic UI distribution.  The library is based on direct DOM manipulation and I don't see much benefit to rewriting the entire framework in favor of using more obtuse vanilla APIs with, inevitably, our own jQuery-like wrapper.  There seems to be a holy war against jQuery though I find little pragmatism to the many of its arguments.

I [removed jQuery entirely](https://react.semantic-ui.com/introduction) from [Semantic-UI-React](https://github.com/Semantic-Org/Semantic-UI-React) as React manipulates its own virtual DOM and wants full solitary control of actual DOM updates.  Because of this, direct DOM manipulation (whether vanilla API or through a wrapper like jQuery) are not serious options.

Hope the extra context helps.

**Comment by RobertFischer** [comment]
On thing that Elm has is a concept of "keyed nodes", which goes a long way to keeping the virtual DOM diffing algorithm from making a mess when the real DOM is hijacked.

The project that I would like to leverage is Quill. It's a very good, very powerful, very extensible Rich Text Editor, and I'd rather not rewrite the whole thing in GHCJS when it exists just fine.

**Comment by steshaw** [comment]
There are a _lot_ of "off the shelf" components that it would be great to be able to integrate with the "virtual-dom way". Examples: code editors (like CodeMirror), rich text editors (like Quill), charts (like d3), date pickers etc! It's a real need (if a temporary one). I'm surprised there isn't better integration between "virtual dom" libraries and "dom manipulation" libraries via some kind of FFI incantations 🔮! Maybe there is now, the world moves fast 😄 !

As an example of CodeMirror, it seems that the Elm guys found a way to integrate that:

- https://github.com/GabiGrin/elm-codemirror

Which leads to:

- http://sgillis.github.io/posts/2016-03-25-highcharts-integration-in-elm.html
- https://github.com/sgillis/elm-highcharts

There's also a [React wrapper for CodeMirror](https://github.com/JedWatson/react-codemirror).

I hope those links are helpful. They could be out of date.

I'm also curious if "virtual dom" components from "other worlds" (such as React or Elm) are easily integrated with Miso. For instance, I discovered that the [Atom](https://atom.io/) editor from GitHub has/had [a React-based editor component](http://blog.atom.io/2014/07/02/moving-atom-to-react.html) which might make an alternative to CodeMirror.

**Comment by dmjio** [comment]
Solved by @FPtje in #362 cc @saurabhnanda
